internal/api/api_test.go

```diff
@@ -0,0 +1,386 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	mock_api "github.com/mattermost/awat/internal/mocks/api"
+	"github.com/mattermost/awat/model"
+)
+
+func TestTranslations(t *testing.T) {
+	logger := MakeLogger(t)
+	mockController := gomock.NewController(t)
+	store := mock_api.NewMockStore(mockController)
+	router := mux.NewRouter()
+	Register(router, &Context{
+		Store:  store,
+		Logger: logger,
+	})
+	ts := httptest.NewServer(router)
+	defer ts.Close()
+
+	t.Run("unknown translation", func(t *testing.T) {
+		store.EXPECT().
+			GetTranslation("bogusID").
+			Return(nil, nil).
+			Times(1)
+
+		resp, err := http.Get(fmt.Sprintf("%s/translation/bogusID", ts.URL))
+		require.NoError(t, err)
+		require.Equal(t, http.StatusNotFound, resp.StatusCode)
+	})
+
+	t.Run("fetch a translation successfully", func(t *testing.T) {
+		translationID := model.NewID()
+		store.EXPECT().
+			GetTranslation(translationID).
+			Return(&model.Translation{ID: translationID}, nil).
+			Times(1)
+
+		resp, err := http.Get(fmt.Sprintf("%s/translation/%s", ts.URL, translationID))
+		require.NoError(t, err)
+		assert.Equal(t, http.StatusOK, resp.StatusCode)
+
+		translation, err := model.NewTranslationStatusFromReader(resp.Body)
+		require.NoError(t, err)
+		assert.Equal(t, translationID, translation.ID)
+	})
+
+	t.Run("encounter an error from the db", func(t *testing.T) {
+		translationID := model.NewID()
+		store.EXPECT().
+			GetTranslation(translationID).
+			Return(nil, errors.New("problem talking to database")).
+			Times(1)
+
+		resp, err := http.Get(fmt.Sprintf("%s/translation/%s", ts.URL, translationID))
+		require.NoError(t, err)
+		assert.Equal(t, http.StatusInternalServerError, resp.StatusCode)
+	})
+
+	t.Run("fetch all translations", func(t *testing.T) {
+		translationID := model.NewID()
+		store.EXPECT().
+			GetAllTranslations().
+			Return([]*model.Translation{
+				{ID: translationID},
+			}, nil).
+			Times(1)
+
+		resp, err := http.Get(fmt.Sprintf("%s/translations", ts.URL))
+		require.NoError(t, err)
+		assert.Equal(t, http.StatusOK, resp.StatusCode)
+
+		translations, err := model.NewTranslationStatusListFromReader(resp.Body)
+		require.NoError(t, err)
+		assert.Equal(t, 1, len(translations))
+		assert.Equal(t, translationID, translations[0].ID)
+	})
+
+	t.Run("start a new translation", func(t *testing.T) {
+		store.EXPECT().
+			StoreTranslation(
+				// a more specific expectation could be applied here, but it
+				// doesn't seem worth the time to define a Matcher and get it
+				// all working just to ignore the nondeterministic ID that's
+				// passed to this function because the ID is generated at
+				// runtime
+				gomock.Any(),
+			).
+			Return(nil).
+			Times(1)
+
+		resp, err := http.Post(fmt.Sprintf("%s/translate", ts.URL), "application/json",
+			strings.NewReader(
+				`{"Type": "slack", "InstallationID": "installationID", "Archive": "foo.zip", "Team": "teamname"}`,
+			))
+		require.NoError(t, err)
```

Comment:
I am just curious, is there any particular reason you chose not to use client methods to make API calls?

Reply:
I thought about it but decided maybe it was better to test one thing at a time? I don't know, maybe that was the wrong decision as I currently don't have tests for the client :joy:

Reply:
So I thought about changing these to use the client and decided that it was just as much work & added more value to just copy-paste the same routines & modify them as a separate test suite, so that this test file can check API return codes & raw HTTP responses directly, and then `client_test.go` can test client behavior, so that's what I've done.